android/src/main/kotlin/co/doneservices/callkeep/IncomingCallActivity.kt

```diff
@@ -0,0 +1,299 @@
+                    finishAndRemoveTask()
+                } else {
+                    finish()
+                }
+            }
+        }
+    }
+
+    private var endedCallKeepBroadcastReceiver = EndedCallKeepBroadcastReceiver()
+
+    private lateinit var llBackground: LinearLayout
+
+    private lateinit var tvCallerName: TextView
+    private lateinit var tvCallHeader: TextView
+    private lateinit var ivLogo: ImageView
+
+    private lateinit var btnAnswer: Button
+
+    private lateinit var btnDecline: Button
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1) {
+            window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
+            setTurnScreenOn(true)
+            setShowWhenLocked(true)
+        } else {
+            window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
+            window.addFlags(WindowManager.LayoutParams.FLAG_TURN_SCREEN_ON)
+            window.addFlags(WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED)
+            window.addFlags(WindowManager.LayoutParams.FLAG_DISMISS_KEYGUARD)
+        }
+        transparentStatusAndNavigation()
+        setContentView(R.layout.activity_call_incoming)
+        initView()
+        incomingData(intent)
+        registerReceiver(
+                endedCallKeepBroadcastReceiver,
+                IntentFilter(ACTION_ENDED_CALL_INCOMING)
+        )
+    }
+
+    private fun wakeLockRequest(duration: Long) {
+
+        val pm = applicationContext.getSystemService(POWER_SERVICE) as PowerManager
+        val wakeLock = pm.newWakeLock(
+                PowerManager.SCREEN_BRIGHT_WAKE_LOCK or PowerManager.FULL_WAKE_LOCK or PowerManager.ACQUIRE_CAUSES_WAKEUP,
+                "CallKeep:PowerManager"
+        )
+        wakeLock.acquire(duration)
+    }
+
+    private fun transparentStatusAndNavigation() {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT && Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
+            setWindowFlag(
+                    WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS
+                            or WindowManager.LayoutParams.FLAG_TRANSLUCENT_NAVIGATION, true
+            )
+        }
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
+            window.decorView.systemUiVisibility = (View.SYSTEM_UI_FLAG_LAYOUT_STABLE
+                    or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
+                    or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION)
+        }
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+            setWindowFlag(
+                    (WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS
+                            or WindowManager.LayoutParams.FLAG_TRANSLUCENT_NAVIGATION), false
+            )
+            window.statusBarColor = Color.TRANSPARENT
+            window.navigationBarColor = Color.TRANSPARENT
+        }
```

Comment:
This is horrible. Aren't there any APIs that roll all these build checks for you?

Reply:
Sadly not, Android land seems very ugly when it comes down to API checks 😅